pkg/storage/db.go

```diff
@@ -0,0 +1,278 @@
+			isJoin = true
+		}
+		err := db.insertIPAddress(ctx, val, tx)
+		if err != nil {
+			return err
+		}
+		err = db.insertNetworkChangeEvent(ctx, timestamp, false, isJoin, resourceID, val, nil, tx)
+		if err != nil {
+			return err
+		}
+	}
+
+	for _, val := range changes.PublicIPAddresses {
+		isJoin := false
+		if changes.ChangeType == "ADDED" {
+			isJoin = true
+		}
+		err := db.insertIPAddress(ctx, val, tx)
+		if err != nil {
+			return err
+		}
+		// yeah, a nested for loop to establish the relationship of every public IP to every hostname
+		for _, hostname := range changes.Hostnames {
+			err := db.insertNetworkChangeEvent(ctx, timestamp, true, isJoin, resourceID, val, &hostname, tx)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func (db *DB) insertIPAddress(ctx context.Context, ipAddress string, tx *sql.Tx) error {
+	// this lib won't give back the last INSERTed row ID, so we don't bother with `RETURNING ...`
+	// See https://stackoverflow.com/questions/34708509/how-to-use-returning-with-on-conflict-in-postgresql
+	if _, err := tx.ExecContext(ctx, fmt.Sprintf(`INSERT INTO %s VALUES ($1) ON CONFLICT DO NOTHING`, tableAWSIPS), ipAddress); err != nil { // nolint
+		return err
+	}
+
+	return nil
+}
+
+func (db *DB) insertHostname(ctx context.Context, hostname string, tx *sql.Tx) error {
+	// this lib won't give back the last INSERTed row ID, so we don't bother with `RETURNING ...`
+	// See https://stackoverflow.com/questions/34708509/how-to-use-returning-with-on-conflict-in-postgresql
+	if _, err := tx.ExecContext(ctx, fmt.Sprintf(`INSERT INTO %s VALUES ($1) ON CONFLICT DO NOTHING`, tableAWSHostnames), hostname); err != nil { // nolint
+		return err
+	}
+
+	return nil
+}
+
+func (db *DB) insertNetworkChangeEvent(ctx context.Context, timestamp time.Time, isPublic bool, isJoin bool, resourceID string, ipAddress string, hostname *string, tx *sql.Tx) error {
+	// Postgres does not auto-create partition tables, so we do it.  We're using 3-month intervals (quarters)
+	monthInterval := 3
+	year := timestamp.Year()
+	fromMonth := (int((timestamp.Month())-1)/monthInterval)*monthInterval + 1 // get the interval of the year we're in, then use the first month of that quarter
+	fromDay := 1
+	toMonth := (int((timestamp.Month())-1)/monthInterval)*monthInterval + monthInterval        // get the interval of the year we're in, then use the last month of that interval
+	toDay := time.Date(year, time.Month(toMonth+1), 0, 0, 0, 0, 0, timestamp.Location()).Day() // time.Date will normalize that toMonth + 1 and 0 day shenanigans
+	from := fmt.Sprintf(`%d-%02d-%02d`, year, fromMonth, fromDay)
+	to := fmt.Sprintf(`%d-%02d-%02d`, year, toMonth, toDay)
+	partitionTableNameSuffix := fmt.Sprintf(`%d_%02dto%02d`, year, fromMonth, toMonth)
+	tableName := fmt.Sprintf(`%s_%s`, tableAWSEventsIPSHostnames, partitionTableNameSuffix)
+
+	if _, err := tx.ExecContext(ctx, fmt.Sprintf(`CREATE TABLE IF NOT EXISTS %s PARTITION OF %s FOR `+ // nolint
```

Comment:
Copying over comments that were lost in the change:
I don't have any data around this but my naive assumption would be that this is an expensive check to run on every insert. Additionally, when this actually results in table creation then my naive expectation is that this is likely an order of magnitude or more slower than an insert especially when factoring in the index creation. Would it be possible to automate this on a schedule instead?

Reply:
Not sure.  It _is_ in a transaction, so not sure if it's really expensive or not.  That said, I _really_ don't like that it's here on every run INSERT.  There are several mechanisms to optimize.  One I was thinking of was just an in-memory array of "known-to-already-exist tables" so we don't bother with that `CREATE`.  That in-memory array could arguably be considered "state", but I don't think it really matters in this case.  I'll impl it, and you guys can comment.

Reply:
In the interest of time and size of this PR, this could potentially be something that we ticket out and address after we merge this code in. We don't _have_ to do it that way. It's only a suggestion.

Reply:
I'm OK with splitting it out to reduce the PR size. In case we continue it here:

The table cache will still result in a cache miss for the first request to each running instance as well as the first request to each running instance every time the clock changes to a new partition-able segment of time.

For the occasions where a new table _is_ created, it's not clear to me how the transaction will affect the query. For example, if we have 5 concurrent writes occurring on each of our 3 running instances then will all 3 instances block until the table create is finished or will two of them fail the transaction? If they block then we become subject to queue based latency as our landing rate for new writes should remain predictable during that time but all writes become enqueued. Alternatively, how would this kind of `DDL in a transaction` work if someone is running this system on a multi-primary configuration of postgresql?

I'd feel much more comfortable if there were separate path we could execute on-demand and in advanced to perform the DDL needed to create a new partition table. This would enable operators to more strictly control this process as well as help mitigate much of the risk of performing dynamic DDL.

Reply:
The thing I might worry about is the timestamp indicated by the AWS config change report being currently processed.  I _think_ we can safely assume that all reports we get have accurate near-real-time timestamps, but we should be careful that we pre-create the partition tables if we're not creating them inline as I've first coded it.  Also, we'd want to code for if/when we need to retrieve or replay anything from cold/old storage.

I'll open up a performance ticket; we need to also avoid constantly running unnecessary INSERTs into the aws_resources table too.

Reply:
Opened SECD-482